SetAuth support for enabling/disabling device authentication

filmgarage opened the pull request:
## Summary

This PR adds methods to enable and disable authentication on Shelly devices for both Gen1 (BlockDevice) and Gen2+ (RpcDevice).

**Type: new-feature**

## Motivation

Currently, aioshelly can authenticate with devices that have authentication enabled, but there's no way to programmatically enable or disable authentication on devices. This is useful for:

- Security automation (enabling auth on all devices)
- Initial device setup workflows
- Password management tools

## Changes

### Gen1 Devices (BlockDevice)

Gen1 devices use a simple HTTP endpoint for authentication management.

```python
# Enable authentication
await device.set_auth("admin", "password")

# Disable authentication
await device.disable_auth()
```

### Gen2+ Devices (RpcDevice)

Gen2+ devices use the `Shelly.SetAuth` RPC method with SHA-256 digest authentication.

```python
# Enable authentication (username is always 'admin')
await device.set_auth("password")

# Disable authentication
await device.disable_auth()
```

## Implementation Details

### Gen1 (BlockDevice)

Uses HTTP GET request to `/settings/login`:
- Enable: `GET /settings/login?enabled=1&username={user}&password={pass}`
- Disable: `GET /settings/login?enabled=0`

Key differences from Gen2+:
- Username can be anything (not limited to "admin")
- Password is sent in plaintext over HTTP

### Gen2+ (RpcDevice)

Authentication Enable Flow:
1. Get device ID from `self.shelly["id"]`
2. Calculate HA1 hash: `SHA256("admin:{device_id}:{password}")`
3. Call `Shelly.SetAuth` with user, realm, and ha1

Authentication Disable Flow:
1. Get device ID from `self.shelly["id"]`
2. Call `Shelly.SetAuth` with user, realm, and `ha1=None`

## Testing

- ✅ Unit tests added for both Gen1 and Gen2+ implementations
- ✅ Manually tested on real Gen1 device (SHSW-1)
- ✅ Manually tested on real Gen2+ device

## References

- [[Shelly Gen1 API - /settings/login](https://shelly-api-docs.shelly.cloud/gen1/#settings-login)](https://shelly-api-docs.shelly.cloud/gen1/#settings-login)
- [[Shelly Gen2 Authentication Documentation](https://shelly-api-docs.shelly.cloud/gen2/General/Authentication/)](https://shelly-api-docs.shelly.cloud/gen2/General/Authentication/)

## Checklist

- [x] Code follows project style guidelines
- [x] Tests added for new functionality
- [x] Tested on real devices

filmgarage committed:
Add methods for authentication set and disable

filmgarage committed:
Add methods for authentication set and disable

filmgarage committed:
Test for authentication on black devices

filmgarage committed:
Test for authentication on rpc devices

filmgarage committed:
Delete tests/rpc_device/test_set_auth.py

filmgarage committed:
Test scripts for set and disable authentication

filmgarage committed:
Delete tests/block_device/test_block_device_auth.py

filmgarage committed:
Test scripts to set and disable authentication

filmgarage committed:
Test for pc device authentication

filmgarage committed:
Test for block device authentication

filmgarage committed:
Delete tests/rpc_device/test_rpc_disable_auth.py

filmgarage committed:
Delete tests/rpc_device/test_rpc_enable_auth.py

filmgarage committed:
Delete tests/block_device/test_block_disable_auth.py

filmgarage committed:
Delete tests/block_device/test_block_enable_auth.py

filmgarage committed:
Merge branch 'main' into main

bieniu commented on tests/block_device/test_block_device_auth.py:
```diff
@@ -0,0 +1,186 @@
+    return device
+
+
+class TestBlockDeviceSetAuth:
```
We do not create separate classes for test cases, please adapt your tests to the current structure.

filmgarage committed:
Simplify docstrings for authentication methods

Removed detailed docstrings for set_auth and disable_auth methods, retaining only brief descriptions.

filmgarage committed:
Simplify docstrings for authentication methods

Removed detailed docstrings for set_auth and disable_auth methods, retaining only brief descriptions.

filmgarage committed:
Implement tests for RpcDevice authentication methods

Added tests for set_auth and disable_auth methods in RpcDevice.

filmgarage committed:
Add tests for BlockDevice authentication methods

filmgarage committed:
Delete tests/block_device/test_block_device_auth.py

filmgarage committed:
Delete tests/rpc_device/test_set_auth.py

filmgarage committed:
Merge branch 'main' into main

filmgarage committed:
Merge branch 'main' into main

chemelli74 commented on aioshelly/rpc_device/device.py:
```diff
@@ -549,6 +549,27 @@ async def switch_set(self, id_: int, value: bool) -> None:
 
+    async def set_auth(self, password: str) -> dict[str, Any]:
+        """Enable authentication on the device."""
+        from .wsrpc import hex_hash
```
Please use pre-commit before committing so you get all the errors ;-)

filmgarage committed:
Replace AsyncMock with MagicMock in tests

filmgarage committed:
Replace MagicMock with AsyncMock in tests

Add AsyncMock to import as well

filmgarage committed:
Merge branch 'main' into main

filmgarage committed:
Combine methods as single set_auth

filmgarage committed:
Combine methods as single set_auth

Updated set_auth method to allow disabling authentication by passing None as the password. Removed the disable_auth method as its functionality is now integrated into set_auth.

filmgarage committed:
Remove whitespace in device.py

filmgarage committed:
import hex_hash

Shorten Docstring

filmgarage committed:
Fix import statement by adding missing comma

filmgarage committed:
Merge branch 'main' into main

filmgarage committed:
Fix import formatting in device.py

filmgarage committed:
Refactor authentication tests to use set_auth method

filmgarage committed:
Rename test for disabling authentication method

filmgarage committed:
Fix test assertions in test_device.py

thecode commented on aioshelly/block_device/device.py:
```diff
@@ -421,6 +421,21 @@ async def trigger_reboot(self) -> None:
+                raise ValueError("username and password required when enabling auth")
+            params = {"enabled": 1, "username": username, "password": password}
+        else:
+            params = {"enabled": 0}
```
According to [Shelly API](https://github.com/home-assistant/core/issues/155884#issuecomment-3498882445) this should be a Boolean with `True` or `False`

filmgarage committed:
Change authentication parameters to use booleans

Following the exact Shelly API

filmgarage committed:
Rename parameter 'enabled' to 'enable' in set_auth

filmgarage committed:
Refactor set_auth to enable/disable authentication

Make more consistent using the enable parameter

filmgarage committed:
Modify set_auth tests to accept boolean values

Follow logo of changed parameters in the method

filmgarage commented on aioshelly/block_device/device.py:
```diff
@@ -421,6 +421,21 @@ async def trigger_reboot(self) -> None:
 
+    async def set_auth(
+        self,
+        enabled: bool,
```
I changed the parameter, not the dictionary keys... Or would I need to change these as well?

filmgarage commented on aioshelly/rpc_device/device.py:
```diff
@@ -549,6 +549,15 @@ async def switch_set(self, id_: int, value: bool) -> None:
         }
         await self.call_rpc("Switch.Set", params=params)
 
+    async def set_auth(self, password: str | None = None) -> dict[str, Any]:
```
That makes sense, more consistent. Tests have also been changed to respect the Bool

chemelli74 commented on tests/block_device/test_device.py:
```diff
@@ -65,3 +65,51 @@ async def test_block_device_ip_address_property_guard(
+    block_device.http_request.assert_called_once_with(
+        "get",
+        "settings/login",
+        {"enabled": 1, "username": "admin", "password": "password123"},
```
```suggestion
        {"enabled": True, "username": "admin", "password": "password123"},
```

filmgarage committed:
Update tests/block_device/test_device.py

Co-authored-by: Simone Chemelli <simone.chemelli@gmail.com>

filmgarage committed:
Update tests/block_device/test_device.py

Co-authored-by: Simone Chemelli <simone.chemelli@gmail.com>

filmgarage committed:
Update aioshelly/block_device/device.py

Co-authored-by: Simone Chemelli <simone.chemelli@gmail.com>

filmgarage committed:
Set unprotected flag and validate string lengths in set_auth

Set unprotected flag to False when disabling authentication. The user will receive a warning on the device web page. Added validation for username and password length in set_auth method.

filmgarage committed:
Modify block device tests and request parameters

Updated HTTP request parameters and added tests for username and password length validation.

filmgarage commented on aioshelly/block_device/device.py:
```diff
@@ -421,6 +421,19 @@ async def trigger_reboot(self) -> None:
+        if enable:
+            if username is None or password is None:
+                raise ValueError("username and password required when enabling auth")
+        params = {"enabled": enable, "username": username, "password": password}
```
Done. I added a validation check for the user and password lengths as well...

chemelli74 replied:
> I think it makes sense to set `unprotected` to `False` if we disable authorization.

If it's the user choice, should be True not False ;-)

filmgarage committed:
Merge branch 'home-assistant-libs:main' into main

filmgarage committed:
remove comment

filmgarage committed:
Refactor authentication length checks with constants

filmgarage committed:
Fix test assertions for block device authentication
